feat: add CloseVector as VectorStore provider

jo32 opened the pull request:
Hi,

This is a pull request that adds support for [CloseVector](https://closevector.getmegaportal.com/), a cross-platform vector database that can currently run on both the browser and Node.js. For instance, you can create your index in Node.js and then load/query it in the browser. 

I've also implemented `saveToCloud` and `loadFromCloud` functions, which allow users to load and save the index to/from the CloudVector CDN.

**Regarding Vitest Integration**

`closevector-web` is a web-based package that requires the DOM and IndexedDB APIs. I attempted to implement Jest tests as shown below:

```closevector_web.test.ts
/**
 * @jest-environment jsdom
 */

import { beforeAll, expect, test } from "@jest/globals";
import { CloseVectorWeb } from "../closevector_web.js";
import { Document } from "../../document.js";
import { FakeEmbeddings } from "../../embeddings/fake.js";

import "fake-indexeddb/auto";

// remaining code
```

However, I encountered the following error:

```error.md
FAIL  src/vectorstores/tests/closevector_web.test.ts
  ● Test suite failed to run

    Jest encountered an unexpected token

    Jest failed to parse a file. This happens e.g. when your code or its dependencies use non-standard JavaScript syntax, or when Jest is not configured to support such syntax.

    Out of the box Jest supports Babel, which will be used to transform your files into valid JS based on your Babel configuration.

    By default "node_modules" folder is ignored by transformers.

    Here's what you can do:
     • If you are trying to use ECMAScript Modules, see https://jestjs.io/docs/ecmascript-modules for how to enable it.
     • If you are trying to use TypeScript, see https://jestjs.io/docs/getting-started#using-typescript
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/configuration
    For information about custom transformations, see:
    https://jestjs.io/docs/code-transformation

    Details:

    /Users/***/langchainjs/node_modules/uuid/dist/esm-browser/index.js:1
    ({"Object.<anonymous>":function(module,exports,require,__dirname,__filename,jest){export { default as v1 } from './v1.js';
                                                                                      ^^^^^^

    SyntaxError: Unexpected token 'export'

      at Runtime.createScriptFromCode (../node_modules/jest-runtime/build/index.js:1495:14)

Test Suites: 1 failed, 1 passed, 2 total
Tests:       2 passed, 2 total
Snapshots:   0 total
Time:        0.641 s, estimated 1 s
Ran all test suites.
```

I attempted to resolve this error, but the solution required numerous workarounds beyond my expertise. Consequently, I integrated Vitest to test the browser-based test cases that follow the **.vitest.ts pattern. Vitest also boasts impressive performance. If this is not acceptable, please let me know, and I will remove the related test cases and dependencies.

jacoblee93 commented:
Very cool! Will review on Monday.

jacoblee93 commented on docs/extras/modules/data_connection/vectorstores/integrations/closevector.mdx:
```diff
@@ -0,0 +1,51 @@
+available on both browser and Node.js
+:::
+
+[CloseVector](https://closevector.getmegaportal.com/) is a cross-platform vector database that can be currently ran on both browser and Node.js. For example, you can create your index on Node.js and then load/query it on browser. For more information, please visit [CloseVector Docs](https://closevector-docs.getmegaportal.com/).
```
s/that can be currently ran on both browser and Node.js/that can run in both the browser and Node.

jo32 replied:
thanks, fixed that

jacoblee93 commented on examples/src/indexes/vector_stores/closevector.ts:
```diff
@@ -0,0 +1,17 @@
+// If you want to import the browser version, use the following line instead:
+// import { CloseVectorWeb } from "langchain/vectorstores/closevector_node";
```
Change to `closevector_web`

jacoblee93 commented on langchain/scripts/create-entrypoints.js:
```diff
@@ -77,6 +77,8 @@ const entrypoints = {
   "vectorstores/elasticsearch": "vectorstores/elasticsearch",
   "vectorstores/memory": "vectorstores/memory",
   "vectorstores/cloudflare_vectorize": "vectorstores/cloudflare_vectorize",
+  "vectorstores/closevector_web": "vectorstores/closevector_web",
```
You might prefer `vectorstores/closevector/web`? Up to you

jo32 replied:
Great suggestions! I have adjusted the structure of the code to make it clearer.

jacoblee93 replied:
You'll also need to move the paths to match under `src/`

jo32 replied:
Yes, references have been updated, including codes in examples and docs. yarn lint, yarn build, and yarn docs have been tested.

jacoblee93 commented on langchain/src/vectorstores/closevector_node.ts:
```diff
@@ -0,0 +1,286 @@
+export class CloseVectorNode extends SaveableVectorStore {
+  declare FilterType: (doc: Document) => boolean;
+
+  _instance?: CloseVectorHNSWNode;
```
Why `this._instance` and `this.instance`?

jo32 replied:
I was thinking about making the setter private to prevent the instance from being changed from the outside.

jacoblee93 commented on langchain/src/vectorstores/closevector_node.ts:
```diff
@@ -0,0 +1,286 @@
+        metadata: x[0].metadata || {},
+      },
+      x[1],
+    ]) as [Document<Record<string, unknown>>, number][];
```
Is cast necessary? Could we just create a `Document` from the `pageContent` and `metadata` and do `as const` instead?

jo32 replied:
Yes, you are correct. I have adjusted it according to your suggestion. Thank you!

jacoblee93 commented on langchain/src/vectorstores/closevector_web.ts:
```diff
@@ -0,0 +1,276 @@
+import {
```
Similar feedback to above

jacoblee93 replied:
I also wonder if it'd be possible to share more code between the two implementations

jo32 replied:
also fixed this. :)

jacoblee93 commented on langchain/src/vectorstores/closevector_web.ts:
```diff
@@ -0,0 +1,276 @@
+   * @param directory The directory to which to save the vector store.
+   * @returns A Promise that resolves when the vector store has been saved.
+   */
+  async save(directory: string): Promise<void> {
```
What is `directory` in a web context?

jo32 replied:
IndexedDB is used to mock the file system. So, in a web context, a directory is more or less like a key saved in IndexedDB. I have also added comments in the code.

jacoblee93 commented:
Looks great overall!

re: `vitest`, would really like to avoid having two different test runners. Would something like this help for mocks?

https://github.com/jakearchibald/idb/issues/165#issuecomment-931682104

Or some other env setup?

https://github.com/langchain-ai/langchainjs/blob/main/langchain/jest.env.cjs#L3

Otherwise, I don't particularly like Jest and would be open to switching everything over to `vitest` if it makes testing in web environments easier generally - do you have any thoughts on comparing the two?

jo32 commented:
> Looks great overall!
> 
> re: `vitest`, would really like to avoid having two different test runners. Would something like this help for mocks?
> 
> [jakearchibald/idb#165 (comment)](https://github.com/jakearchibald/idb/issues/165#issuecomment-931682104)
> 
> Or some other env setup?
> 
> https://github.com/langchain-ai/langchainjs/blob/main/langchain/jest.env.cjs#L3
> 
> Otherwise, I don't particularly like Jest and would be open to switching everything over to `vitest` if it makes testing in web environments easier generally - do you have any thoughts on comparing the two?

@jacoblee93 I noticed that SWC is already used in the Jest config file. So, in terms of performance, I think there are no significant differences. I might spend extra time trying to figure out what's wrong first and attempt to make it work with Jest.

jo32 commented:
@jacoblee93

It seems that when using jsdom or happy-dom as the testing environment in Jest, it encounters an issue: https://github.com/jestjs/jest/issues/13739#issuecomment-1517190965.

I have tried all the solutions recommended here https://github.com/jestjs/jest/issues/13739#issuecomment-1517190965 and none have solved the problem.

In the end, I might have to rewrite all the test cases in vitest style and replace Jest with vitest. In this case, this commit will be too large and not focused enough. Therefore, I removed the web-related test cases and rewrote the node ones back in Jest style for the time being. However, the web-related functions are tested with vitest.

I might submit another issue, replacing Jest with vitest, if I figure out that vitest is the only option for now.

Also, all the other comments have been resolved and the code has been submitted.

jacoblee93 commented:
Ah bummer :( Thanks, will look as soon as I can!

jacoblee93 commented on langchain/src/vectorstores/closevector/node.ts:
```diff
@@ -0,0 +1,175 @@
+  static async imports(): Promise<{
+    HierarchicalNSW: typeof HierarchicalNSWT;
+  }> {
+    return await CloseVectorHNSWNode.imports();
```
No need for return `await` here?

Also, I think no need for conditional import?

jo32 replied:
@jacoblee93

True! I have removed "await" in the latest code.

If possible, I might prefer to reserve this method to allow developers to debug the inner instance and also align it with the closevector-web implementation, which can be used to preload the JS resources on the web.

jacoblee93 commented on langchain/src/vectorstores/closevector/common.ts:
```diff
@@ -0,0 +1,227 @@
+import type { CloseVectorHNSWNode } from "closevector-node";
+import type { CloseVectorHNSWWeb } from "closevector-web";
```
This will cause import issues unfortunately :(

You'll need to separate imports out into the individual entrypoints.

jo32 replied:
@jacoblee93 

Ah! Sorry for not noticing that. I have deleted the import from closevector-node or closevector-web in common.ts; instead, I imported an type from a package named closevector-common, which is a package shared by closevector-node and closevector-web in latest code.

jo32 commented:
@jacoblee93 Thank you for your feedback. The most recent comments have been addressed. I would greatly appreciate any additional suggestions you may have.

jacoblee93 commented on langchain/src/vectorstores/closevector/common.ts:
```diff
@@ -0,0 +1,226 @@
+  async saveToCloud(
+    _options: Record<string, unknown>
+  ) {
+    throw new Error("not implemented");
```
Shouldn't these be abstract methods?

jo32 replied:
Oh, yes, I have made it abstract.

jacoblee93 commented on langchain/src/vectorstores/closevector/common.ts:
```diff
@@ -0,0 +1,226 @@
+    throw new Error("not implemented");
+  }
+
+  static async imports(): Promise<{
```
What's the purpose of this method? Looks like it's not called?

jo32 replied:
I was assuming LangChain might be used in some non-TypeScript environments that may not have solid type checking and tests. A class method in common.ts may be called directly. Thus, a runtime error is added to hint at such cases.

However, I forgot to check if there are any practices like this in LangChain. I have removed those methods in common.ts to adhere to the conventions.

jacoblee93 commented on docs/snippets/modules/data_connection/vectorstores/get_started.mdx:
```diff
@@ -116,3 +116,4 @@ Here's a quick guide to help you pick the right vector store for your use case:
 - If you are already utilizing SingleStore, or if you find yourself in need of a distributed, high-performance database, you might want to consider the [SingleStore](./integrations/singlestore) vector store.
 - If you are looking for an online MPP (Massively Parallel Processing) data warehousing service, you might want to consider the [AnalyticDB](./integrations/analyticdb) vector store.
 - If you're in search of a cost-effective vector database that allows run vector search with SQL, look no further than [MyScale](./integrations/myscale).
+- If you're in search of a vector database that can loaded on both browser and server side, look no further than [CloseVector](./integrations/closevector). It aims to running vector database on cross-platform.
```
Some typos here - how about:

`- If you're in search of a vector database that you can load from both the browser and server side, check out [CloseVector](./integrations/closevector). It's a vector database that aims to be cross-platform.`

jo32 replied:
Great idea! I edited the docs in the latest commit.

jacoblee93 commented:
Think we're almost there!

jacoblee93 commented:
Could you also run `yarn format`?

jo32 commented:
@jacoblee93 thank you. The code has been formatted and submitted again. It's a great honor to have the opportunity to contribute to a mature and rapidly developing open-source project like LangChain.

jacoblee93 commented:
Thank you!

We were dealing with some CI timeout issues and docs refactors, so I'll need to merge in main but should be good :)
